indexer-balances/src/balance_client.rs

```diff
@@ -0,0 +1,114 @@
+use std::str::FromStr;
+
+use crate::models::balance_changes::NearBalanceEvent;
+use crate::models::SqlxMethods;
+use async_trait::async_trait;
+use near_lake_framework::near_indexer_primitives;
+
+#[async_trait]
+pub trait BalanceClient {
+    async fn get_balance_before_block(
+        &self,
+        account_id: &near_indexer_primitives::types::AccountId,
+        block_header: &near_indexer_primitives::views::BlockHeaderView,
+    ) -> anyhow::Result<crate::BalanceDetails>;
+}
+
+pub struct JsonRpcBalanceClient {
+    json_rpc_client: near_jsonrpc_client::JsonRpcClient,
+}
+
+impl JsonRpcBalanceClient {
+    pub fn new(json_rpc_client: near_jsonrpc_client::JsonRpcClient) -> Self {
+        Self { json_rpc_client }
+    }
+}
+
+#[async_trait]
+impl BalanceClient for JsonRpcBalanceClient {
+    async fn get_balance_before_block(
+        &self,
+        account_id: &near_indexer_primitives::types::AccountId,
+        block_header: &near_indexer_primitives::views::BlockHeaderView,
+    ) -> anyhow::Result<crate::BalanceDetails> {
+        let query = near_jsonrpc_client::methods::query::RpcQueryRequest {
+            block_reference: near_primitives::types::BlockReference::BlockId(
+                near_primitives::types::BlockId::Hash(block_header.prev_hash),
+            ),
+            request: near_primitives::views::QueryRequest::ViewAccount {
+                account_id: account_id.clone(),
+            },
+        };
+
+        let account_response = self.json_rpc_client.call(query).await;
+
+        if let Err(err) = account_response {
+            return match err.handler_error() {
+                Some(near_jsonrpc_primitives::types::query::RpcQueryError::UnknownAccount {
+                    ..
+                }) => Ok(crate::BalanceDetails {
+                    non_staked: 0,
+                    staked: 0,
+                }),
+                _ => Err(err.into()),
+            };
+        }
+
+        let response_kind = account_response.unwrap().kind;
```

Comment:
Let's avoid unwrap(). Here it could be done by using `match` earlier

Reply:
Woops, good catch! Changed to:
```rust
let response_kind = account_response?.kind;
```